x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/TopSnippets.java

```diff
@@ -0,0 +1,260 @@
+        this.field = field;
+        this.query = query;
+        this.options = options;
+    }
+
+    public TopSnippets(StreamInput in) throws IOException {
+        this(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteable(Expression.class),
+            in.readOptionalNamedWriteable(Expression.class)
+        );
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field);
+        out.writeNamedWriteable(query);
+        out.writeOptionalNamedWriteable(options);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isString(field(), sourceText(), FIRST).and(() -> isString(query(), sourceText(), SECOND))
+            .and(() -> resolve(options(), source(), THIRD, ALLOWED_OPTIONS));
+    }
+
+    @Override
+    public boolean foldable() {
+        return field().foldable() && query().foldable() && (options() == null || options().foldable());
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new TopSnippets(
+            source(),
+            newChildren.get(0), // field
+            newChildren.get(1), // query
+            newChildren.size() > 2 ? newChildren.get(2) : null // options
+        );
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, TopSnippets::new, field, query, options);
+    }
+
+    Expression field() {
+        return field;
+    }
+
+    Expression query() {
+        return query;
+    }
+
+    Expression options() {
+        return options;
+    }
+
+    private int numSnippets() {
+        return extractIntegerOption(NUM_SNIPPETS, DEFAULT_NUM_SNIPPETS);
+    }
+
+    private int numWords() {
+        return extractIntegerOption(NUM_WORDS, DEFAULT_WORD_SIZE);
+    }
+
+    private int extractIntegerOption(String option, int defaultValue) {
+        if (options == null) {
+            return defaultValue;
+        }
+
+        MapExpression optionsMap = (MapExpression) options;
+        Expression expr = optionsMap.keyFoldedMap().get(option);
+        if (expr == null) {
+            return defaultValue;
+        }
+        Object value = expr.fold(FoldContext.small());
+        return value != null ? ((Number) value).intValue() : defaultValue;
+    }
+
+    @Evaluator(extraName = "BytesRef")
+    static void process(
+        BytesRefBlock.Builder builder,
+        BytesRef str,
+        BytesRef query,
```

Comment:
Is query a constant, or can it be different for every row? If it's a constant let's make it `@Fixed`